File: VB/SecuredExportExample.Module/SecurityObjects/ExtendedSecurityRole.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.ComponentModel

Imports DevExpress.Xpo
Imports DevExpress.Data.Filtering

Imports DevExpress.ExpressApp
Imports DevExpress.Persistent.Base
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.Persistent.Validation
Imports DevExpress.ExpressApp.Security

Namespace SecuredExportExample.Module.SecurityObjects
    <DefaultClassOptions, ImageName("BO_Role")> _
    Public Class ExtendedSecurityRole
        Inherits SecurityRole
        Public Sub New(ByVal session As Session)
            MyBase.New(session)
        End Sub
        Private Function FindExportPermissionData() As ExportPermissionData
            For Each permissionData As PermissionData In PersistentPermissions
                Dim exportPermissionData As ExportPermissionData = TryCast(permissionData, ExportPermissionData)
                If exportPermissionData IsNot Nothing Then
                    Return exportPermissionData
                End If
            Next permissionData
            Return Nothing
        End Function
        <NonPersistent> _
        Public Property CanExport() As Boolean
            Get
                Dim exportPermissionData As ExportPermissionData = FindExportPermissionData()
                Return exportPermissionData IsNot Nothing AndAlso exportPermissionData.CanExport
            End Get
            Set(ByVal value As Boolean)
                If (Not IsLoading) Then
                    Dim exportPermissionData As ExportPermissionData = FindExportPermissionData()
                    If value Then
                        If exportPermissionData Is Nothing Then
                            exportPermissionData = New ExportPermissionData(Session)
                            PersistentPermissions.Add(exportPermissionData)
                        End If
                        exportPermissionData.CanExport = True
                    Else
                        If exportPermissionData IsNot Nothing Then
                            exportPermissionData.CanExport = False

                        End If
                    End If
                End If
            End Set
        End Property
    End Class
End Namespace

File: CS/SecuredExportExample.Module/SecurityObjects/ExtendedSecurityRole.cs
using System;
using System.ComponentModel;

using DevExpress.Xpo;
using DevExpress.Data.Filtering;

using DevExpress.ExpressApp;
using DevExpress.Persistent.Base;
using DevExpress.Persistent.BaseImpl;
using DevExpress.Persistent.Validation;
using DevExpress.ExpressApp.Security;

namespace SecuredExportExample.Module.SecurityObjects {
    [DefaultClassOptions]
    [ImageName("BO_Role")]
    public class ExtendedSecurityRole : SecurityRole {
        public ExtendedSecurityRole(Session session) : base(session) { }
        private ExportPermissionData FindExportPermissionData() {
            foreach (PermissionData permissionData in PersistentPermissions) {
                ExportPermissionData exportPermissionData = permissionData as ExportPermissionData;
                if (exportPermissionData != null) {
                    return exportPermissionData;
                }
            }
            return null;
        }
        [NonPersistent]
        public bool CanExport {
            get {
                ExportPermissionData exportPermissionData = FindExportPermissionData();
                return exportPermissionData != null && exportPermissionData.CanExport;
            }
            set {
                if (!IsLoading) {
                    ExportPermissionData exportPermissionData = FindExportPermissionData();
                    if (value) {
                        if (exportPermissionData == null) {
                            exportPermissionData = new ExportPermissionData(Session);
                            PersistentPermissions.Add(exportPermissionData);
                        }
                        exportPermissionData.CanExport = true;
                    }
                    else {
                        if (exportPermissionData != null) {
                            exportPermissionData.CanExport = false;
                            
                        }
                    }
                }
            }
        }
    }
}
